src/main/java/edu/wpi/grip/ui/pipeline/AddSourceView.java

```diff
@@ -22,19 +35,65 @@
+     * freezing up the GUI.
+     */
+    private final Executor loadSourceExecutor = Executors.newSingleThreadExecutor(runnable -> {
+        final Thread thread = new Thread(runnable);
+        thread.setDaemon(true);
+        return thread;
+    });
+
     public AddSourceView(EventBus eventBus) {
         this.eventBus = eventBus;
 
         this.setFillHeight(true);
 
         addButton("Add\nImage", getClass().getResource("/edu/wpi/grip/ui/icons/add-image.png"), mouseEvent -> {
-            // TODO: create an ImageFileSource and send a SourceAddedEvent
-            System.out.println(mouseEvent);
+            // Show a file picker so the user can open one or more images from disk
+            final FileChooser fileChooser = new FileChooser();
+            fileChooser.setTitle("Open an image");
+
+            final List<File> imageFiles = fileChooser.showOpenMultipleDialog(this.getScene().getWindow());
+
+            if (imageFiles == null) return;
+
+            // Add a new source for each image .
+            imageFiles.forEach(file -> {
+                this.loadSourceExecutor.execute(() -> {
+                    try {
+                        final ImageFileSource source = new ImageFileSource(eventBus, file.toURI().toURL());
+                        eventBus.post(new SourceAddedEvent(source));
+                    } catch (MalformedURLException e) {
+                        e.printStackTrace();
+                    }
+                });
+            });
         });
 
         addButton("Add\nWebcam", getClass().getResource("/edu/wpi/grip/ui/icons/add-webcam.png"), mouseEvent -> {
-            // TODO: create a WebcamSource and send a SourceAddedEvent
-            System.out.println(mouseEvent);
+            final Parent root = this.getScene().getRoot();
+
+            // Show a dialog for the user to pick a camera index
+            final Dialog<ButtonType> dialog = new Dialog<>();
+            final Spinner<Integer> cameraIndex = new Spinner<Integer>(0, Integer.MAX_VALUE, 0);
+
+            dialog.setTitle("Add Webcam");
+            dialog.setHeaderText("Choose a camera");
+            dialog.setContentText("index");
+            dialog.getDialogPane().setContent(cameraIndex);
+            dialog.getDialogPane().getButtonTypes().addAll(ButtonType.CANCEL, ButtonType.OK);
+            dialog.getDialogPane().setStyle(root.getStyle());
+            dialog.getDialogPane().getStylesheets().addAll(root.getStylesheets());
+
+            // If the user clicks OK, add a new webcam source
+            dialog.showAndWait().filter(Predicate.isEqual(ButtonType.OK)).ifPresent(result -> {
+                this.loadSourceExecutor.execute(() -> {
+                    final WebcamSource source = new WebcamSource(eventBus, cameraIndex.getValue());
+                    eventBus.post(new SourceAddedEvent(source));
+                });
+            });
```

Comment:
There should probably be `create` and `destroy` methods or something like that in `Source`, since this problem will also affect network sources